README.md
# SYNOPSIS

[![NPM Package](https://img.shields.io/npm/v/srlp.svg?style=flat-square)](https://www.npmjs.org/package/srlp)
[![Build Status](https://img.shields.io/travis/sophonjs/srlp.svg?branch=master&style=flat-square)](https://travis-ci.org/sophonjs/srlp)
[![Coverage Status](https://img.shields.io/coveralls/sophonjs/srlp.svg?style=flat-square)](https://coveralls.io/r/sophonjs/srlp)
[![Gitter](https://img.shields.io/gitter/room/sophon/sophonjs-lib.svg?style=flat-square)](https://gitter.im/sophon/sophonjs-lib) or #sophonjs on freenode

[Recursive Length](https://octonion.institute/susy-go/wiki/wiki/SRLP) Prefix Encoding for node.js.

## INSTALL

`npm install srlp`

install with `-g` if you want to use the cli.

## USAGE

```javascript
var SRLP = require('srlp')
var assert = require('assert')

var nestedList = [[], [[]], [[], [[]]]]
var encoded = SRLP.encode(nestedList)
var decoded = SRLP.decode(encoded)
assert.deepEqual(nestedList, decoded)
```

## API

`srlp.encode(plain)` - SRLP encodes an `Array`, `Buffer` or `String` and returns a `Buffer`.

`srlp.decode(encoded, [skipRemainderCheck=false])` - Decodes an SRLP encoded `Buffer`, `Array` or `String` and returns a `Buffer` or an `Array` of `Buffers`. If `skipRemainderCheck` is enabled, `srlp` will just decode the first srlp sequence in the buffer. By default, it would throw an error if there are more bytes in Buffer than used by srlp sequence.

## CLI

`srlp decode <hex string>`  
`srlp encode <json String>`

## TESTS

Test uses mocha. To run tests and linting: `npm test`. To auto fix linting problems use: `npm run test:fix`.

## CODE COVERAGE

Install dev dependencies
`npm install`

Run
`npm run coverage`

The results are at
`coverage/lcov-report/index.html`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
(modification: no type change headlines) and this project adheres to
[Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [v2.2.3] - 2019-03-19

- More robust `Array` type checking in `SRLP.encode()` function,
  PR [#70](https://octonion.institute/susy-js/srlp/pull/70)
- Library now throws an error when trying to encode negative integer
  values (e.g. `SRLP.encode(-1)`),
  PR [#71](https://octonion.institute/susy-js/srlp/pull/70)

[v2.2.3]: https://octonion.institute/susy-js/srlp/compare/2.2.1...v2.2.3

## [v2.2.2] - 2019-01-15

- Added `bn.js` dependency to fix module resolution bug when require the module after
  clean install, PR [#64](https://octonion.institute/susy-js/srlp/pull/64)
- Use local version of official tests, fixed testing issue, PR [#66](https://octonion.institute/susy-js/srlp/pull/66)

[v2.2.2]: https://octonion.institute/susy-js/srlp/compare/2.2.1...v2.2.2

## [v2.2.1] - 2018-12-20

- Fixed a bug introduced in `v2.2.0` causing the CLI `bin/srlp` command not to
  work, see PR [#60](https://octonion.institute/susy-js/srlp/pull/60)
- Additional exports of types used by `decode` and `encode`
  (PR [#59](https://octonion.institute/susy-js/srlp/pull/59)):
  - `Input`: input type for `encode()`
  - `Dictionary` and `List`: interfaces for possible `Input` values
  - `Decoded`: interface for `decode()` return type
- Additional test structure and new integration tests for distribution and
  `bin/srlp` CLI command, see PR [#57](https://octonion.institute/susy-js/srlp/pull/57)

[v2.2.1]: https://octonion.institute/susy-js/srlp/compare/2.2.0...v2.2.1

## [v2.2.0] - 2018-12-17

[DEPRECATED] Please update to `v2.2.1`, release contains a broken CLI `bin/srlp`
command!

First `TypeScript` based release of `SophonJS` libraries, thanks @krzkaczor,
@GrandSchtroumpf, @whymarrh, @holgerd77 for the great work on this!

This release doesn't introduce any main new features but will serve as a basis
for further `TypeScript` transitions coming along with greater type safety
for `SophonJS` libraries. If you are developing in `TypeScript` you can
further already benefit from the `SRLP` type declarations published from now on
along new releases.

See PR [#37](https://octonion.institute/susy-js/srlp/pull/37) and subsequent PRs merged
towards the associated `typescript` branch to get an overview on the changes.
The release also comes along with the introduction of a new repo
[sophonjs-config](https://octonion.institute/susy-js/sophonjs-config) centralizing
configuration for `SophonJS` libraries on `TypeScript` itself as well as
linting, formatting and testing.

This release passes all existing unit tests and other checks. If you nevertheless
experience problems please report on the `SophonJS`
[Gitter](https://gitter.im/sophon/sophonjs) channel.

Other changes:

- Added `LICENSE` file for `MPL2.0`, see PR [#31](https://octonion.institute/susy-js/srlp/pull/31)

[v2.2.0]: https://octonion.institute/susy-js/srlp/compare/2.1.0...v2.2.0

## [v2.1.0] - 2018-06-28

- Updated supported Node versions, PR [#13](https://octonion.institute/susy-js/srlp/pull/13)
- Switched to `safe-buffer` for backwards compatibility, PR [#18](https://octonion.institute/susy-js/srlp/pull/18)
- Increased test coverage, PR [#22](https://octonion.institute/susy-js/srlp/pull/22)
- Example code tweaks, PR [#12](https://octonion.institute/susy-js/srlp/pull/12)
- Fix test runs on Windows, Issue [#7](https://octonion.institute/susy-js/srlp/issues/7)
- Added code coverage, PR [#8](https://octonion.institute/susy-js/srlp/pull/8)

[v2.1.0]: https://octonion.institute/susy-js/srlp/compare/2.0.0...v2.1.0

## [2.0.0] - 2015-09-23

- User `Buffer` values as input for encoding

[2.0.0]: https://octonion.institute/susy-js/srlp/compare/1.1.2...2.0.0

## [1.1.2] - 2015-09-22

- Fix zero encoding

[1.1.2]: https://octonion.institute/susy-js/srlp/compare/1.1.1...1.1.2

## [1.1.1] - 2015-09-21

- Fixes for `bin`

[1.1.1]: https://octonion.institute/susy-js/srlp/compare/1.1.0...1.1.1

## [1.1.0] - 2015-09-21

- Added `getLength()` method
- Added hex prefix stripping (`isHexPrefix()` / `stripHexPrefix()`)
- Code formatting clean-ups

[1.1.0]: https://octonion.institute/susy-js/srlp/compare/1.0.1...1.1.0

## [1.0.1] - 2015-06-27

- Code formatting clean-ups

[1.0.1]: https://octonion.institute/susy-js/srlp/compare/1.0.0...1.0.1

## [1.0.0] - 2015-06-06

- Added check for invalid 0
- Hardened srlp

[1.0.0]: https://octonion.institute/susy-js/srlp/compare/0.0.14...1.0.0

## Older releases:

- [0.0.14](https://octonion.institute/susy-js/srlp/compare/0.0.13...0.0.14) - 2015-03-31
- [0.0.13](https://octonion.institute/susy-js/srlp/compare/0.0.12...0.0.13) - 2015-03-30
- [0.0.12](https://octonion.institute/susy-js/srlp/compare/0.0.11...0.0.12) - 2014-12-26


test/official.spec.ts
import * as assert from 'assert'
import * as SRLP from '../src'
const BN = require('bn.js')

describe('offical tests', function() {
  const officalTests = require('./fixture/srlptest.json').tests

  for (const testName in officalTests) {
    it(`should pass ${testName}`, function(done) {
      let incoming = officalTests[testName].in
      // if we are testing a big number
      if (incoming[0] === '#') {
        const bn = new BN(incoming.slice(1))
        incoming = Buffer.from(bn.toArray())
      }

      const encoded = SRLP.encode(incoming)
      assert.equal('0x' + encoded.toString('hex'), officalTests[testName].out.toLowerCase())
      done()
    })
  }
})


test/dataTypes.spec.ts
import * as assert from 'assert'
import * as SRLP from '../src'
const BN = require('bn.js')

describe('invalid srlps', function() {
  it('should not crash on an invalid srlp', function() {
    // prettier-ignore
    const a = Buffer.from([239, 191, 189, 239, 191, 189, 239, 191, 189, 239, 191, 189, 239, 191, 189, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 239, 191, 189, 29, 239, 191, 189, 77, 239, 191, 189, 239, 191, 189, 239, 191, 189, 93, 122, 239, 191, 189, 239, 191, 189, 239, 191, 189, 103, 239, 191, 189, 239, 191, 189, 239, 191, 189, 26, 239, 191, 189, 18, 69, 27, 239, 191, 189, 239, 191, 189, 116, 19, 239, 191, 189, 239, 191, 189, 66, 239, 191, 189, 64, 212, 147, 71, 239, 191, 189, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 239, 191, 189, 11, 222, 155, 122, 54, 42, 194, 169, 239, 191, 189, 70, 239, 191, 189, 72, 239, 191, 189, 239, 191, 189, 54, 53, 239, 191, 189, 100, 73, 239, 191, 189, 55, 239, 191, 189, 239, 191, 189, 59, 1, 239, 191, 189, 109, 239, 191, 189, 239, 191, 189, 93, 239, 191, 189, 208, 128, 239, 191, 189, 239, 191, 189, 0, 239, 191, 189, 239, 191, 189, 239, 191, 189, 15, 66, 64, 239, 191, 189, 239, 191, 189, 239, 191, 189, 239, 191, 189, 4, 239, 191, 189, 79, 103, 239, 191, 189, 85, 239, 191, 189, 239, 191, 189, 239, 191, 189, 74, 239, 191, 189, 239, 191, 189, 239, 191, 189, 239, 191, 189, 54, 239, 191, 189, 239, 191, 189, 239, 191, 189, 239, 191, 189, 239, 191, 189, 83, 239, 191, 189, 14, 239, 191, 189, 239, 191, 189, 239, 191, 189, 4, 63, 239, 191, 189, 63, 239, 191, 189, 41, 239, 191, 189, 239, 191, 189, 239, 191, 189, 67, 28, 239, 191, 189, 239, 191, 189, 11, 239, 191, 189, 31, 239, 191, 189, 239, 191, 189, 104, 96, 100, 239, 191, 189, 239, 191, 189, 12, 239, 191, 189, 239, 191, 189, 206, 152, 239, 191, 189, 239, 191, 189, 31, 112, 111, 239, 191, 189, 239, 191, 189, 65, 239, 191, 189, 41, 239, 191, 189, 239, 191, 189, 53, 84, 11, 239, 191, 189, 239, 191, 189, 12, 102, 24, 12, 42, 105, 109, 239, 191, 189, 58, 239, 191, 189, 4, 239, 191, 189, 104, 82, 9, 239, 191, 189, 6, 66, 91, 43, 38, 102, 117, 239, 191, 189, 105, 239, 191, 189, 239, 191, 189, 239, 191, 189, 89, 127, 239, 191, 189, 114])
    try {
      SRLP.decode(a)
    } catch (e) {
      // FIXME: check for exception name
      assert(true)
    }
  })
})

describe('SRLP encoding (string):', function() {
  it('should return itself if single byte and less than 0x7f:', function() {
    const encodedSelf = SRLP.encode('a')
    assert.equal(encodedSelf.toString(), 'a')
    assert.equal(SRLP.getLength(encodedSelf), 1)
  })

  it('length of string 0-55 should return (0x80+len(data)) plus data', function() {
    const encodedDog = SRLP.encode('dog')
    assert.equal(4, encodedDog.length)
    assert.equal(SRLP.getLength(encodedDog), 4)
    assert.equal(encodedDog[0], 131)
    assert.equal(encodedDog[1], 100)
    assert.equal(encodedDog[2], 111)
    assert.equal(encodedDog[3], 103)
  })

  it('length of string >55 should return 0xb7+len(len(data)) plus len(data) plus data', function() {
    const encodedLongString = SRLP.encode(
      'zoo255zoo255zzzzzzzzzzzzssssssssssssssssssssssssssssssssssssssssssssss',
    )
    assert.equal(72, encodedLongString.length)
    assert.equal(SRLP.getLength(encodedLongString), 2)
    assert.equal(encodedLongString[0], 184)
    assert.equal(encodedLongString[1], 70)
    assert.equal(encodedLongString[2], 122)
    assert.equal(encodedLongString[3], 111)
    assert.equal(encodedLongString[12], 53)
  })
})

describe('SRLP encoding (list):', function() {
  it('length of list 0-55 should return (0xc0+len(data)) plus data', function() {
    const encodedArrayOfStrings = SRLP.encode(['dog', 'god', 'cat'])
    assert.equal(13, encodedArrayOfStrings.length)
    assert.equal(encodedArrayOfStrings[0], 204)
    assert.equal(encodedArrayOfStrings[1], 131)
    assert.equal(encodedArrayOfStrings[11], 97)
    assert.equal(encodedArrayOfStrings[12], 116)
  })

  // it('length of list >55 should return 0xf7+len(len(data)) plus len(data) plus data', function () {
  //   // need a test case here!
  // })
})

describe('SRLP encoding (BN):', function() {
  it('should encode a BN value', function() {
    const encodedBN = SRLP.encode(new BN(3))
    assert.equal(encodedBN[0], 3)
  })
})

describe('SRLP encoding (integer):', function() {
  it('length of int = 1, less than 0x7f, similar to string', function() {
    const encodedNumber = SRLP.encode(15)
    assert.equal(1, encodedNumber.length)
    assert.equal(encodedNumber[0], 15)
  })

  it('length of int > 55, similar to string', function() {
    const encodedNumber = SRLP.encode(1024)
    assert.equal(3, encodedNumber.length)
    assert.equal(encodedNumber[0], 130)
    assert.equal(encodedNumber[1], 4)
    assert.equal(encodedNumber[2], 0)
  })

  it('it should handle zero', function() {
    assert.equal(SRLP.encode(0).toString('hex'), '80')
  })
})

describe('SRLP decoding (string):', function() {
  it('first byte < 0x7f, return byte itself', function() {
    const decodedStr = SRLP.decode(Buffer.from([97]))
    assert.equal(1, decodedStr.length)
    assert.equal(decodedStr.toString(), 'a')
  })

  it('first byte < 0xb7, data is everything except first byte', function() {
    const decodedStr = SRLP.decode(Buffer.from([131, 100, 111, 103]))
    assert.equal(3, decodedStr.length)
    assert.equal(decodedStr.toString(), 'dog')
  })

  it('array', function() {
    // prettier-ignore
    const decodedBufferArray = SRLP.decode(Buffer.from([204, 131, 100, 111, 103, 131, 103, 111, 100, 131, 99, 97, 116]))
    assert.deepEqual(decodedBufferArray, [
      Buffer.from('dog'),
      Buffer.from('god'),
      Buffer.from('cat'),
    ])
  })
})

describe('SRLP decoding (int):', function() {
  it('first byte < 0x7f, return itself', function() {
    const decodedSmallNum = SRLP.decode(Buffer.from([15]))
    assert.equal(1, decodedSmallNum.length)
    assert.equal(decodedSmallNum[0], 15)
  })

  it('first byte < 0xb7, data is everything except first byte', function() {
    const decodedNum = SRLP.decode(Buffer.from([130, 4, 0]))
    assert.equal(2, decodedNum.length)
    assert.equal(decodedNum.toString('hex'), '0400')
  })
})

describe('strings over 55 bytes long', function() {
  const testString =
    'This function takes in a data, convert it to buffer if not, and a length for recursion'
  const testBuffer = Buffer.from(testString)
  let encoded: Buffer

  it('should encode it', function() {
    encoded = SRLP.encode(testBuffer)
    assert.equal(encoded[0], 184)
    assert.equal(encoded[1], 86)
  })

  it('should decode', function() {
    const decoded = SRLP.decode(encoded)
    assert.equal(decoded.toString(), testString)
  })
})

describe('list over 55 bytes long', function() {
  // prettier-ignore
  const testString = ['This', 'function', 'takes', 'in', 'a', 'data', 'convert', 'it', 'to', 'buffer', 'if', 'not', 'and', 'a', 'length', 'for', 'recursion', 'a1', 'a2', 'a3', 'ia4', 'a5', 'a6', 'a7', 'a8', 'ba9']
  let encoded: Buffer

  it('should encode it', function() {
    encoded = SRLP.encode(testString)
  })

  it('should decode', function() {
    const decodedBuffer = SRLP.decode(encoded)
    const decoded: string[] = []
    for (let i = 0; i < decodedBuffer.length; i++) {
      decoded[i] = decodedBuffer[i].toString()
    }
    assert.deepEqual(decoded, testString)
  })
})

describe('nested lists:', function() {
  // prettier-ignore
  const nestedList = [
    [],
    [
      []
    ],
    [
      [],
      [
        []
      ]
    ]
  ]
  const valueList = [
    [1, 2, 3],
    [
      Buffer.from([4, 5, 6]),
      Buffer.from([7, 8, 9]),
      [Buffer.from([0]), Buffer.from('abcd', 'hex')],
    ],
  ]
  let encoded: Buffer
  it('encode a nested list', function() {
    encoded = SRLP.encode(nestedList)
    assert.deepEqual(encoded, Buffer.from([0xc7, 0xc0, 0xc1, 0xc0, 0xc3, 0xc0, 0xc1, 0xc0]))
  })

  it('should decode a nested list', function() {
    const decoded = SRLP.decode(encoded)
    assert.deepEqual(nestedList, decoded)
  })

  it('should encode a list with values', function() {
    const valueEncoded = SRLP.encode(valueList)
    // prettier-ignore
    assert.deepEqual(valueEncoded, Buffer.from([0xd2, 0xc3, 0x01, 0x02, 0x03, 0xcd, 0x83, 0x04, 0x05, 0x06, 0x83, 0x07, 0x08, 0x09, 0xc4, 0x00, 0x82, 0xab, 0xcd]))
  })
})

describe('typed lists:', function() {
  const valueList = [
    [1, 2, 3],
    [
      new Uint8Array([4, 5, 6]),
      new Uint8Array([7, 8, 9]),
      [new Uint8Array([0]), Buffer.from('abcd', 'hex')],
    ],
  ]

  // equivalent to list of values above
  it('encode a nested list', function() {
    const valueEncoded = SRLP.encode(valueList)
    // prettier-ignore
    assert.deepEqual(valueEncoded, new Buffer([0xd2, 0xc3, 0x01, 0x02, 0x03, 0xcd, 0x83, 0x04, 0x05, 0x06, 0x83, 0x07, 0x08, 0x09, 0xc4, 0x00, 0x82, 0xab, 0xcd]))
  })
})

describe('null values', function() {
  const nestedList = [null]
  let encoded
  it('encode a null array', function() {
    encoded = SRLP.encode(nestedList)
    assert.deepEqual(encoded, Buffer.from([0xc1, 0x80]))
  })

  it('should decode a null value', function() {
    assert.deepEqual(Buffer.from([]), SRLP.decode(Buffer.from('80', 'hex')))
  })
})

describe('zero values', function() {
  let encoded
  it('encode a zero', function() {
    encoded = SRLP.encode(Buffer.from([0]))
    assert.deepEqual(encoded, Buffer.from([0]))
  })

  it('decode a zero', function() {
    const decode = SRLP.decode(Buffer.from([0]))
    assert.deepEqual(decode, Buffer.from([0]))
  })
})

describe('empty values', function() {
  let decoded
  it('decode empty buffer', function() {
    decoded = SRLP.decode(Buffer.from([]))
    assert.deepEqual(decoded, Buffer.from([]))
  })
})

describe('bad values', function() {
  it('wrong encoded a zero', function() {
    const val = Buffer.from(
      'f9005f030182520894b94f5374fce5edbc8e2a8697c15331677e6ebf0b0a801ca098ff921201554726367d2be8c804a7ff89ccf285ebc57dff8ae4c44b9c19ac4aa08887321be575c8095f789dd4c743dfe42c1820f9231f98a962b210e3ac2452a3',
      'hex',
    )
    let result
    try {
      result = SRLP.decode(val)
    } catch (e) {}
    assert.equal(result, undefined)
  })

  it('invalid length', function() {
    const a = Buffer.from(
      'f86081000182520894b94f5374fce5edbc8e2a8697c15331677e6ebf0b0a801ca098ff921201554726367d2be8c804a7ff89ccf285ebc57dff8ae4c44b9c19ac4aa08887321be575c8095f789dd4c743dfe42c1820f9231f98a962b210e3ac2452a3',
      'hex',
    )

    let result
    try {
      result = SRLP.decode(a)
    } catch (e) {}
    assert.equal(result, undefined)
  })

  it('extra data at end', function() {
    const c =
      'f90260f901f9a02a3c692012a15502ba9c39f3aebb36694eed978c74b52e6c0cf210d301dbf325a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347948888f1f195afa192cfee860698584c030f4c9db1a0ef1552a40b7165c3cd773806b9e0c165b75356e0314bf0706f279c729f51e017a0b6c9fd1447d0b414a1f05957927746f58ef5a2ebde17db631d460eaf6a93b18da0bc37d79753ad738a6dac4921e57392f145d8887476de3f783dfa7edae9283e52b90100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008302000001832fefd8825208845509814280a00451dd53d9c09f3cfb627b51d9d80632ed801f6330ee584bffc26caac9b9249f88c7bffe5ebd94cc2ff861f85f800a82c35094095e7baea6a6c7c4c2dfeb977efac326af552d870a801ba098c3a099885a281885f487fd37550de16436e8c47874cd213531b10fe751617fa044b6b81011ce57bffcaf610bf728fb8a7237ad261ea2d937423d78eb9e137076c0ef'

    const a = Buffer.from(c, 'hex')

    let result
    try {
      result = SRLP.decode(a)
    } catch (e) {}
    assert.equal(result, undefined)
  })

  it('extra data at end', function() {
    const c =
      'f9ffffffc260f901f9a02a3c692012a15502ba9c39f3aebb36694eed978c74b52e6c0cf210d301dbf325a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347948888f1f195afa192cfee860698584c030f4c9db1a0ef1552a40b7165c3cd773806b9e0c165b75356e0314bf0706f279c729f51e017a0b6c9fd1447d0b414a1f05957927746f58ef5a2ebde17db631d460eaf6a93b18da0bc37d79753ad738a6dac4921e57392f145d8887476de3f783dfa7edae9283e52b90100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008302000001832fefd8825208845509814280a00451dd53d9c09f3cfb627b51d9d80632ed801f6330ee584bffc26caac9b9249f88c7bffe5ebd94cc2ff861f85f800a82c35094095e7baea6a6c7c4c2dfeb977efac326af552d870a801ba098c3a099885a281885f487fd37550de16436e8c47874cd213531b10fe751617fa044b6b81011ce57bffcaf610bf728fb8a7237ad261ea2d937423d78eb9e137076c0'

    const a = Buffer.from(c, 'hex')

    let result
    try {
      result = SRLP.decode(a)
    } catch (e) {}
    assert.equal(result, undefined)
  })

  it('list length longer than data', function() {
    const c =
      'f9ffffffc260f901f9a02a3c692012a15502ba9c39f3aebb36694eed978c74b52e6c0cf210d301dbf325a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347948888f1f195afa192cfee860698584c030f4c9db1a0ef1552a40b7165c3cd773806b9e0c165b75356e0314bf0706f279c729f51e017a0b6c9fd1447d0b414a1f05957927746f58ef5a2ebde17db631d460eaf6a93b18da0bc37d79753ad738a6dac4921e57392f145d8887476de3f783dfa7edae9283e52b90100000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000008302000001832fefd8825208845509814280a00451dd53d9c09f3cfb627b51d9d80632ed801f6330ee584bffc26caac9b9249f88c7bffe5ebd94cc2ff861f85f800a82c35094095e7baea6a6c7c4c2dfeb977efac326af552d870a801ba098c3a099885a281885f487fd37550de16436e8c47874cd213531b10fe751617fa044b6b81011ce57bffcaf610bf728fb8a7237ad261ea2d937423d78eb9e137076c0'

    const a = Buffer.from(c, 'hex')

    let result
    try {
      result = SRLP.decode(a)
    } catch (e) {}
    assert.equal(result, undefined)
  })
})

describe('hex prefix', function() {
  it('should have the same value', function() {
    const a = SRLP.encode('0x88f')
    const b = SRLP.encode('88f')
    assert.notEqual(a.toString('hex'), b.toString('hex'))
  })
})

describe('recursive typings', function() {
  it('should not throw compilation error', function() {
    type IsType<T, U> = Exclude<T, U> extends never
      ? Exclude<U, T> extends never
        ? true
        : false
      : false
    const assertType = <T, U>(isTrue: IsType<T, U>) => {
      return isTrue
    }
    // tslint:disable-next-line:no-dead-store
    const a = SRLP.encode([[[[[0]]]]])
    assert.ok(assertType<typeof a, Buffer>(true))
  })
})


test/integration.spec.ts
import * as assert from 'assert'
import { exec } from 'child_process'
import * as SRLP from '../dist'
import * as vm from 'vm'

describe('Distribution:', function() {
  it('should be able to execute functionality from distribution build', function() {
    const encodedSelf = SRLP.encode('a')
    assert.equal(encodedSelf.toString(), 'a')
    assert.equal(SRLP.getLength(encodedSelf), 1)
  })
})

describe('CLI command:', function() {
  it('should be able to run CLI command', function() {
    exec('./bin/srlp encode "[ 5 ]"', (_error, stdout, _stderr) => {
      assert.equal(stdout.trim(), 'c105')
    })
  })
})

describe('Cross-frame:', function() {
  it('should be able to encode Arrays across stack frames', function() {
    assert.equal(
      vm.runInNewContext("SRLP.encode(['dog', 'god', 'cat']).toString('hex')", { SRLP }),
      'cc83646f6783676f6483636174',
    )
  })
})


src/index.ts
import BN = require('bn.js')

import { Decoded, Dictionary, Input, List } from './types'

// Types exported outside of this package
export { Decoded, Dictionary, Input, List }

/**
 * SRLP Encoding based on: https://octonion.institute/susy-go/wiki/wiki/%5BEnglish%5D-SRLP
 * This function takes in a data, convert it to buffer if not, and a length for recursion
 * @param input - will be converted to buffer
 * @returns returns buffer of encoded data
 **/
export function encode(input: Input): Buffer {
  if (Array.isArray(input)) {
    const output: Buffer[] = []
    for (let i = 0; i < input.length; i++) {
      output.push(encode(input[i]))
    }
    const buf = Buffer.concat(output)
    return Buffer.concat([encodeLength(buf.length, 192), buf])
  } else {
    const inputBuf = toBuffer(input)
    return inputBuf.length === 1 && inputBuf[0] < 128
      ? inputBuf
      : Buffer.concat([encodeLength(inputBuf.length, 128), inputBuf])
  }
}

/**
 * Parse integers. Check if there is no leading zeros
 * @param v The value to parse
 * @param base The base to parse the integer into
 */
function safeParseInt(v: string, base: number): number {
  if (v.slice(0, 2) === '00') {
    throw new Error('invalid SRLP: extra zeros')
  }

  return parseInt(v, base)
}

function encodeLength(len: number, offset: number): Buffer {
  if (len < 56) {
    return Buffer.from([len + offset])
  } else {
    const hexLength = intToHex(len)
    const lLength = hexLength.length / 2
    const firstByte = intToHex(offset + 55 + lLength)
    return Buffer.from(firstByte + hexLength, 'hex')
  }
}

/**
 * SRLP Decoding based on: {@link https://octonion.institute/susy-go/wiki/wiki/%5BEnglish%5D-SRLP|SRLP}
 * @param input - will be converted to buffer
 * @param stream - Is the input a stream (false by default)
 * @returns - returns decode Array of Buffers containg the original message
 **/
export function decode(input: Buffer, stream?: boolean): Buffer
export function decode(input: Buffer[], stream?: boolean): Buffer[]
export function decode(input: Input, stream?: boolean): Buffer[] | Buffer | Decoded
export function decode(input: Input, stream: boolean = false): Buffer[] | Buffer | Decoded {
  if (!input || (<any>input).length === 0) {
    return Buffer.from([])
  }

  const inputBuffer = toBuffer(input)
  const decoded = _decode(inputBuffer)

  if (stream) {
    return decoded
  }
  if (decoded.remainder.length !== 0) {
    throw new Error('invalid remainder')
  }

  return decoded.data
}

/**
 * Get the length of the SRLP input
 * @param input
 * @returns The length of the input or an empty Buffer if no input
 */
export function getLength(input: Input): Buffer | number {
  if (!input || (<any>input).length === 0) {
    return Buffer.from([])
  }

  const inputBuffer = toBuffer(input)
  const firstByte = inputBuffer[0]

  if (firstByte <= 0x7f) {
    return inputBuffer.length
  } else if (firstByte <= 0xb7) {
    return firstByte - 0x7f
  } else if (firstByte <= 0xbf) {
    return firstByte - 0xb6
  } else if (firstByte <= 0xf7) {
    // a list between  0-55 bytes long
    return firstByte - 0xbf
  } else {
    // a list  over 55 bytes long
    const llength = firstByte - 0xf6
    const length = safeParseInt(inputBuffer.slice(1, llength).toString('hex'), 16)
    return llength + length
  }
}

/** Decode an input with SRLP */
function _decode(input: Buffer): Decoded {
  let length, llength, data, innerRemainder, d
  const decoded = []
  const firstByte = input[0]

  if (firstByte <= 0x7f) {
    // a single byte whose value is in the [0x00, 0x7f] range, that byte is its own SRLP encoding.
    return {
      data: input.slice(0, 1),
      remainder: input.slice(1),
    }
  } else if (firstByte <= 0xb7) {
    // string is 0-55 bytes long. A single byte with value 0x80 plus the length of the string followed by the string
    // The range of the first byte is [0x80, 0xb7]
    length = firstByte - 0x7f

    // set 0x80 null to 0
    if (firstByte === 0x80) {
      data = Buffer.from([])
    } else {
      data = input.slice(1, length)
    }

    if (length === 2 && data[0] < 0x80) {
      throw new Error('invalid srlp encoding: byte must be less 0x80')
    }

    return {
      data: data,
      remainder: input.slice(length),
    }
  } else if (firstByte <= 0xbf) {
    llength = firstByte - 0xb6
    length = safeParseInt(input.slice(1, llength).toString('hex'), 16)
    data = input.slice(llength, length + llength)
    if (data.length < length) {
      throw new Error('invalid SRLP')
    }

    return {
      data: data,
      remainder: input.slice(length + llength),
    }
  } else if (firstByte <= 0xf7) {
    // a list between  0-55 bytes long
    length = firstByte - 0xbf
    innerRemainder = input.slice(1, length)
    while (innerRemainder.length) {
      d = _decode(innerRemainder)
      decoded.push(d.data as Buffer)
      innerRemainder = d.remainder
    }

    return {
      data: decoded,
      remainder: input.slice(length),
    }
  } else {
    // a list  over 55 bytes long
    llength = firstByte - 0xf6
    length = safeParseInt(input.slice(1, llength).toString('hex'), 16)
    const totalLength = llength + length
    if (totalLength > input.length) {
      throw new Error('invalid srlp: total length is larger than the data')
    }

    innerRemainder = input.slice(llength, totalLength)
    if (innerRemainder.length === 0) {
      throw new Error('invalid srlp, List has a invalid length')
    }

    while (innerRemainder.length) {
      d = _decode(innerRemainder)
      decoded.push(d.data as Buffer)
      innerRemainder = d.remainder
    }
    return {
      data: decoded,
      remainder: input.slice(totalLength),
    }
  }
}

/** Check if a string is prefixed by 0x */
function isHexPrefixed(str: string): boolean {
  return str.slice(0, 2) === '0x'
}

/** Removes 0x from a given String */
function stripHexPrefix(str: string): string {
  if (typeof str !== 'string') {
    return str
  }
  return isHexPrefixed(str) ? str.slice(2) : str
}

/** Transform an integer into its hexadecimal value */
function intToHex(integer: number): string {
  if (integer < 0) {
    throw new Error('Invalid integer as argument, must be unsigned!')
  }
  const hex = integer.toString(16)
  return hex.length % 2 ? `0${hex}` : hex
}

/** Pad a string to be even */
function padToEven(a: string): string {
  return a.length % 2 ? `0${a}` : a
}

/** Transform an integer into a Buffer */
function intToBuffer(integer: number): Buffer {
  const hex = intToHex(integer)
  return Buffer.from(hex, 'hex')
}

/** Transform anything into a Buffer */
function toBuffer(v: Input): Buffer {
  if (!Buffer.isBuffer(v)) {
    if (typeof v === 'string') {
      if (isHexPrefixed(v)) {
        return Buffer.from(padToEven(stripHexPrefix(v)), 'hex')
      } else {
        return Buffer.from(v)
      }
    } else if (typeof v === 'number') {
      if (!v) {
        return Buffer.from([])
      } else {
        return intToBuffer(v)
      }
    } else if (v === null || v === undefined) {
      return Buffer.from([])
    } else if (v instanceof Uint8Array) {
      return Buffer.from(v as any)
    } else if (BN.isBN(v)) {
      // converts a BN to a Buffer
      return Buffer.from(v.toArray())
    } else {
      throw new Error('invalid type')
    }
  }
  return v
}
